packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,326 @@
+/**
+ * External dependencies
+ */
+import { createElement, Fragment } from 'react';
+
+let indoc,
+	offset,
+	output,
+	stack,
+	keyIndex;
+
+/**
+ * Matches tags in the localized string
+ *
+ * This is used for extracting the tag pattern groups for parsing the localized
+ * string and along with the map converting it to a react element.
+ *
+ * There are four references extracted using this tokenizer:
+ *
+ * match: Full match of the tag (i.e. <strong>, </strong>, <br/>)
+ * isClosing: If the match has a closing tag (i.e. </)
+ * name: The name portion of the tag (strong, br) (if )
+ * isSelfClosed: Whether the match is for a self-closing tag (i.e. />)
+ *
+ * @type RegExp
+ */
+const tokenizer = /<(\/)?(\w+)\s*(\/)?>/g;
+
+/**
+ * This receives the value for a map element which consists of an element and
+ * it's props and returns a element creator function.
+ *
+ * @param {Array} [ element, props ] The first item in the array is expected to
+ *                                   be a string or a react component. The
+ *                                   second item is expected to be either
+ *                                   undefined or an object.
+ */
+const elementCreator = ( [ element, props = {} ] ) => ( children ) => {
+	if ( typeof element !== 'string' && typeof element !== 'function' ) {
+		throw new Error( `Not a valid element (${ element })` );
+	}
+	props.key = ++keyIndex;
```

Comment:
this is one of those cases where we might be making the linter happy at the expense of the intent of the lint. if we create a random numeric key we both won't help the diffing algorithm know which element is in the right spot but also we risk introducing diffing bugs by saying that two separate things are equal.

this is one of those things that probably is best resolved in a specific spot in the code and not in a general wrapper function.

Reply:
> if we create a random numeric key we both won't help the diffing algorithm know which element is in the right spot but also we risk introducing diffing bugs by saying that two separate things are equal.

Is it truly random though? We're parsing a string which should be consistent so the incrementation will be consistent through each parse right?  Even if in the future language is flipped through context or something, the parent component would likely be re-rendered anyways so there shouldn't be diffing issues?

> this is one of those things that probably is best resolved in a specific spot in the code and not in a general wrapper function.

Any ideas where/how here?

Reply:
> We're parsing a string which should be consistent so the incrementation will be consistent through each parse right?

It's random and obscured by the fact that we're mixing two different levels of "rendering" here.

```js
<div>
	{ hasPosts ? translate(
		'<Icon /> Posts', { Icon: <Icon name="posts" /> }
	) : translate(
		'<Icon /> No Posts', { Icon: <Icon name="missing" /> }
	) }
</div>
```

This probably isn't the right situation, but it gets close. The `div` controls the context for where `key` becomes important and we could re-render the translation differently without changing the parent context. This is where bugs can slip in.

> Any ideas where/how here?

Leave it for the parent component to use? We don't need to deal with it here. If we can't build real test cases that show it fails you can ignore all of this, but here's what I'd consider…

```js
translate( '<Icon /> Show the things!<End />', {
	Icon: <Icon key="icon" />,
	End: <Icon key="suffix" />,
} )
```

Could be different in every situation

Reply:
addressed in d7796d9